File: sdd-ext/engine/internal/sdd/propose.go
package sdd

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// maxSeedTitleLen bounds the title derived from a free-text description so the
// frontmatter title (and the resulting decision slug) stays reasonable.
const maxSeedTitleLen = 72

// SeedProposal writes sdd/proposal.md as an in-review skeleton seeded from a
// natural-language description, without invoking any model. It is the graceful
// fallback for `grok-sdd propose` when no provider is configured: the human (or a
// later agent turn) fills in the Context/Acceptance sections, then approves.
//
// It runs Scaffold first so a missing sdd/ is created, and refuses to clobber a
// proposal that is already in review — surfacing that instead of overwriting.
// Returns the proposal path relative to root.
func SeedProposal(root, description string, now time.Time) (string, error) {
	description = strings.TrimSpace(description)
	if description == "" {
		return "", fmt.Errorf("a proposal description is required")
	}
	if _, _, err := Scaffold(root); err != nil {
		return "", fmt.Errorf("scaffold sdd/: %w", err)
	}

	base := filepath.Join(root, DirName)
	proposalPath := filepath.Join(base, "proposal.md")
	if raw, err := os.ReadFile(proposalPath); err == nil {
		fm, _ := splitFrontmatter(string(raw))
		if status := strings.TrimSpace(fm["status"]); status != "" && status != "empty" {
			return "", fmt.Errorf("sdd/proposal.md is already in review (%q); approve or clear it before drafting a new one", status)
		}
	}

	title := deriveTitle(description)
	var doc strings.Builder
	doc.WriteString("---\n")
	doc.WriteString("type: Proposal\n")
	doc.WriteString("title: ")
	doc.WriteString(title)
	doc.WriteString("\n")
	doc.WriteString("description: \n")
	doc.WriteString("tags: []\n")
	doc.WriteString("status: in-review\n")
	doc.WriteString("timestamp: ")
	doc.WriteString(now.UTC().Format(time.RFC3339))
	doc.WriteString("\n")
	doc.WriteString("---\n\n")
	doc.WriteString("# Proposal\n\n")
	doc.WriteString(description)
	doc.WriteString("\n\n")
	doc.WriteString("_Seeded skeleton — this file already exists; edit it in place to expand the what/why above, then fill Context and Acceptance below._\n\n")
	doc.WriteString("# Context\n\n")
	doc.WriteString("The forces, constraints, and alternatives considered.\n\n")
	doc.WriteString("# Acceptance\n\n")
	doc.WriteString("What must be true for this proposal to be approved and promoted to `decisions/NNN-name.md`.\n")

	relPath := filepath.Join(DirName, "proposal.md")
	if err := os.WriteFile(proposalPath, []byte(doc.String()), 0o644); err != nil {
		return "", err
	}
	return relPath, nil
}

// ProposalBranchName derives the feature branch a proposal — and all of its
// later work — lives on, from the same title slug the approved decision will
// carry: sdd/prop-<slug>. One branch per proposal, so the doc, its approval, and
// the implementation land in a single PR rather than scattered across main.
func ProposalBranchName(description string) string {
	slug := slugify(deriveTitle(description))
	if slug == "" {
		slug = "proposal"
	}
	return "sdd/prop-" + slug
}

// deriveTitle turns the first sentence/line of a description into a concise
// frontmatter title, trimmed to maxSeedTitleLen on a word boundary.
func deriveTitle(description string) string {
	first := description
	if idx := strings.IndexAny(first, ".\n"); idx > 0 {
		first = first[:idx]
	}
	first = strings.TrimSpace(strings.Join(strings.Fields(first), " "))
	if first == "" {
		return "Untitled proposal"
	}
	if len(first) <= maxSeedTitleLen {
		return first
	}
	clipped := first[:maxSeedTitleLen]
	if sp := strings.LastIndex(clipped, " "); sp > 0 {
		clipped = clipped[:sp]
	}
	return strings.TrimSpace(clipped) + "…"
}

File: crates/common/xai-grok-sdd/src/propose.rs
//! Seeding a proposal from a natural-language description, without invoking any
//! model — the graceful fallback for `grok-sdd propose`. Ported from
//! `internal/sdd/propose.go`.

use crate::scaffold::{DIR_NAME, scaffold};
use crate::util::{slugify, split_frontmatter};
use anyhow::{Context, bail};
use chrono::{DateTime, Utc};
use std::fs;
use std::path::Path;

/// Bounds the title derived from a free-text description so the frontmatter title
/// (and the resulting decision slug) stays reasonable.
const MAX_SEED_TITLE_LEN: usize = 72;

/// Writes `sdd/proposal.md` as an in-review skeleton seeded from a description.
/// Runs [`scaffold`] first so a missing `sdd/` is created, and refuses to clobber
/// a proposal already in review. Returns the proposal path relative to root.
pub fn seed_proposal(root: &Path, description: &str, now: DateTime<Utc>) -> anyhow::Result<String> {
    let description = description.trim();
    if description.is_empty() {
        bail!("a proposal description is required");
    }
    scaffold(root).context("scaffold sdd/")?;

    let base = root.join(DIR_NAME);
    let proposal_path = base.join("proposal.md");
    if let Ok(raw) = fs::read_to_string(&proposal_path) {
        let (fm, _) = split_frontmatter(&raw);
        let status = fm.get("status").map(|s| s.trim()).unwrap_or("");
        if !status.is_empty() && status != "empty" {
            bail!(
                "sdd/proposal.md is already in review ({status:?}); approve or clear it before drafting a new one"
            );
        }
    }

    let title = derive_title(description);
    let doc = format!(
        "---\ntype: Proposal\ntitle: {title}\ndescription: \ntags: []\nstatus: in-review\ntimestamp: {ts}\n---\n\n\
         # Proposal\n\n{description}\n\n\
         _Seeded skeleton — this file already exists; edit it in place to expand the what/why above, then fill Context and Acceptance below._\n\n\
         # Context\n\nThe forces, constraints, and alternatives considered.\n\n\
         # Acceptance\n\nWhat must be true for this proposal to be approved and promoted to `decisions/NNN-name.md`.\n",
        ts = now.format("%Y-%m-%dT%H:%M:%SZ"),
    );
    fs::write(&proposal_path, doc)?;
    Ok(format!("{DIR_NAME}/proposal.md"))
}

/// Derives the feature branch a proposal — and all of its later work — lives on,
/// from the same title slug the approved decision will carry: `sdd/prop-<slug>`.
pub fn proposal_branch_name(description: &str) -> String {
    let mut slug = slugify(&derive_title(description));
    if slug.is_empty() {
        slug = "proposal".to_string();
    }
    format!("sdd/prop-{slug}")
}

/// Turns the first sentence/line of a description into a concise frontmatter
/// title, trimmed to [`MAX_SEED_TITLE_LEN`] bytes on a word boundary.
fn derive_title(description: &str) -> String {
    let mut first = description;
    if let Some(idx) = description.find(['.', '\n'])
        && idx > 0
    {
        first = &description[..idx];
    }
    let first: String = first.split_whitespace().collect::<Vec<_>>().join(" ");
    if first.is_empty() {
        return "Untitled proposal".to_string();
    }
    if first.len() <= MAX_SEED_TITLE_LEN {
        return first;
    }
    let budget = MAX_SEED_TITLE_LEN.min(first.len());
    // Prefer a word boundary within the byte budget; else floor to a char boundary.
    let clip = match first.as_bytes()[..budget].iter().rposition(|&b| b == b' ') {
        Some(sp) if sp > 0 => sp,
        _ => {
            let mut b = budget;
            while b > 0 && !first.is_char_boundary(b) {
                b -= 1;
            }
            b
        }
    };
    format!("{}…", first[..clip].trim())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn fixed() -> DateTime<Utc> {
        DateTime::parse_from_rfc3339("2026-07-25T10:00:00Z")
            .unwrap()
            .with_timezone(&Utc)
    }

    #[test]
    fn seed_proposal_writes_in_review_skeleton() {
        let tmp = tempfile::tempdir().unwrap();
        let root = tmp.path();
        let rel = seed_proposal(
            root,
            "Build a POS for a single pollería. Offline-first cash flow.",
            fixed(),
        )
        .unwrap();
        assert_eq!(rel, "sdd/proposal.md");
        let body = fs::read_to_string(root.join(&rel)).unwrap();
        for want in [
            "status: in-review",
            "title: Build a POS for a single pollería",
            "# Proposal",
            "Offline-first cash flow",
            "# Acceptance",
        ] {
            assert!(body.contains(want), "proposal missing {want:?}:\n{body}");
        }
    }

    #[test]
    fn seed_proposal_scaffolds_when_missing() {
        let tmp = tempfile::tempdir().unwrap();
        let root = tmp.path();
        seed_proposal(root, "First idea", fixed()).unwrap();
        assert!(
            root.join("sdd/index.md").exists(),
            "expected sdd/ scaffolded"
        );
    }

    #[test]
    fn seed_proposal_refuses_to_clobber_active_proposal() {
        let tmp = tempfile::tempdir().unwrap();
        let root = tmp.path();
        seed_proposal(root, "First idea", fixed()).unwrap();
        let err = seed_proposal(root, "Second idea", fixed()).unwrap_err();
        assert!(
            err.to_string().contains("already in review"),
            "error = {err}"
        );
    }

    #[test]
    fn seed_proposal_requires_description() {
        let tmp = tempfile::tempdir().unwrap();
        assert!(seed_proposal(tmp.path(), "   ", fixed()).is_err());
    }

    #[test]
    fn derive_title_truncates_on_word_boundary() {
        let long = "word ".repeat(40);
        let got = derive_title(&long);
        assert!(
            got.len() <= MAX_SEED_TITLE_LEN + "…".len(),
            "title too long ({}): {got:?}",
            got.len()
        );
        assert!(got.ends_with('…'), "expected ellipsis suffix, got {got:?}");
    }

    #[test]
    fn proposal_branch_name_cases() {
        let cases = [
            (
                "SaaS multi-tenant para pollerías en Perú. Offline.",
                "sdd/prop-saas-multi-tenant-para-polleras-en-per",
            ),
            (
                "Architecture with TypeScript and InstantDB",
                "sdd/prop-architecture-with-typescript-and-instantdb",
            ),
            ("   ", "sdd/prop-untitled-proposal"),
            (
                "Fix vuelto rounding bug",
                "sdd/prop-fix-vuelto-rounding-bug",
            ),
        ];
        for (desc, want) in cases {
            assert_eq!(
                proposal_branch_name(desc),
                want,
                "ProposalBranchName({desc:?})"
            );
        }
    }
}
